src/lib.rs

```diff
@@ -430,3 +430,15 @@ pub enum InlineAsmDialect {
     #[llvm_variant(LLVMInlineAsmDialectIntel)]
     Intel,
 }
+
+
+/// Exposes LLVM internal references to the outside.
+#[cfg(feature = "internal-getters")]
+pub trait LLVMReference<T> {
```

Comment:
Since all types can only resolve to exactly one type, this should be an associated type and not a generic:

```rust
pub trait LLVMReference {
    type Ref;
    unsafe fn get_ref(&self) -> Self::Ref;
}
```

Reply:
I changed the trait as requested, the only problem I had there is that the implementation for `AsValueRef` and `AsTypeRef` will require `dyn` as the trait will conflict otherwise.
https://stackoverflow.com/questions/39159226/conflicting-implementations-of-trait-in-rust

Reply:
I'm not sure I follow, what exactly is conflicting?

Reply:
I was trying to implement `LLVMReference` for all types implementing `AsValueRef` and all types implementing `AsTypeRef`
So my implementation looks like this
```rust
//values/mod.rs
#[cfg(feature="internal-getters")]
impl<T : AsValueRef> LLVMReference for T {
    type Ref = LLVMValueRef;
    unsafe fn get_ref(&self) -> Self::Ref {
       self.as_value_ref() 
    }
}
//types/mod.rs
#[cfg(feature="internal-getters")]
impl<T : AsTypeRef> LLVMReference for T {
    type Ref  = LLVMTypeRef;
    unsafe fn get_ref(&self) -> Self::Ref {
       self.as_type_ref() 
    }
}
```
This does not compile because of a conflicting implementation
```
   --> src/values/mod.rs:227:1
    |
227 | impl<T : AsValueRef> LLVMReference for T {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation
    |
   ::: src/types/mod.rs:207:1
    |
207 | impl<T : AsTypeRef> LLVMReference for T {
    | --------------------------------------- first implementation here
```
If I understand correctly, this is because there is no guarantee that types implementing `AsTypeRef` will not at a future point also implement `AsValueRef`
My solution was to use `dyn`. Maybe I'm missing something?

Reply:
hmm, I see. I didn't know you could impl for a dyn trait. I wonder if it works as expected 🤔

Reply:
It seems from a quick [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=10cd3e70b9a8a687e0291dce7dc11bf2) try that this does not work as intended.

I'll look for options to keep this without generics. Some [option ](https://geo-ant.github.io/blog/2021/mutually-exclusive-traits-rust/)I found was to try and emulate mutually exclusive traits by using a wrapper trait. But I'm still not sure if that's the best option

Reply:
It does work, but you have to explicitly cast to the dyn trait with is pretty ugly: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f49b5ce1409f4977235fbf9f201d3fd6

Reply:
I'll add some tests and use generics then. I was thinking if I limit the generics only to the AsRef implementations it would then keep the associated types everywhere else, which makes more sense for 1-1 mappings.
I'll give it a shot, and if I can't manage it I will revert to generics everywhere